### PR TITLE
fix: refine data entry app locked messages for [DHIS2-4674]

### PR DESCRIPTION
**Pull request by tomzemp**
This PR refines message for data sets to provide more detail than the current general purpose `Data set is locked`. See https://jira.dhis2.org/browse/DHIS2-4674 for background.

We have used the following messages:

| Situation | Message in UI |
| ------- | ------- |
| Date is past expiry days for period | Data entry has concluded |
| Date is outside of defined data input periods | Data input period has closed |
| Data is approved | Data has been approved and cannot be modified |
| Other* | Data set is locked |

\* General purpose message is left in case locked definition on backend (from DataSetService) is updated to cover locking reasons beyond those above.

GetDataValuesForDataSetAction has been modified to split current locked value into locked and approved values to provide necessary detail to frontend for determining the appropriate message.

### CODE REVIEW

**Review by jimgrace — APPROVED** [review]
Looks good to me. Thanks!

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java [hunk 1]
```diff
@@ -335,8 +346,10 @@ public String execute()
                 lastUpdatedBy = registration.getLastUpdatedBy();
             }
 
-            locked = dataSetService.isLocked( currentUser, dataSet, period, organisationUnit, attributeOptionCombo,
-                null );
+            locked = dataSetService.isLocked( currentUser, dataSet, period, organisationUnit, null );
+            approved = dataApprovalService.isApproved( dataSet.getWorkflow(), period, organisationUnit,
```
Comment (larshelge) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java:350]: Many thanks for the nice contribution. One issue here is that the check for `dataApprovalService.isApproved` is also currently done within `dataSetService.isLocked`. This means we are doing the check twice, which is not great from a performance perspective, nor from a readability perspective.

Instead I suggest that we change the return type of dataSetService.isLocked to an enum, and rename the method to something like `getLockStatus`. The enum could have values such as OPEN, LOCKED, APPROVED, etc. That way the semantics is cleaner and performance is better.

Comment (jimgrace) [reply dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java:350]: @larshelge Actually, this overload of `dataSetService.isLocked` doesn't call `dataApprovalService.isApproved`. It's called by the longer overload of `isLocked` that also calls `isApproved`.

I agree that the enum would be less confusing all around and could also be used for SMS.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataset/DefaultDataSetService.java [hunk 2]
```diff
@@ -266,6 +267,23 @@ public int getLockExceptionCount()
...
+    @Override
+    @Transactional( readOnly = true )
+    public LockStatus getLockStatus( User user, DataSet dataSet, Period period, OrganisationUnit organisationUnit,
+        CategoryOptionCombo attributeOptionCombo, Date now )
+    {
+        if ( dataApprovalService.isApproved( dataSet.getWorkflow(), period, organisationUnit, attributeOptionCombo ) )
+        {
+            return LockStatus.APPROVED;
+        }
+        if ( isLocked( user, dataSet, period, organisationUnit, now ) )
```
Comment (jimgrace) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataset/DefaultDataSetService.java:279]: Insert blank line before

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataset/DefaultDataSetService.java [hunk 3]
```diff
@@ -266,6 +267,23 @@ public int getLockExceptionCount()
...
+        {
+            return LockStatus.LOCKED;
+        }
+        return LockStatus.OPEN;
```
Comment (jimgrace) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataset/DefaultDataSetService.java:283]: Insert blank line before

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataset/DefaultDataSetService.java [hunk 4]
```diff
@@ -266,6 +267,23 @@ public int getLockExceptionCount()
...
+    }
+
+
```
Comment (jimgrace) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataset/DefaultDataSetService.java:286]: Remove one of the two blank lines

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java [hunk 5]
```diff
@@ -126,6 +128,9 @@ public void setCurrentUserService( CurrentUserService currentUserService )
     @Autowired
     private InputUtils inputUtils;
 
+    @Autowired
+    private DataApprovalService dataApprovalService;
+
```
Comment (jimgrace) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java:133]: Can remove this (and the import). No longer referenced

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java [hunk 6]
```diff
@@ -335,8 +340,9 @@ public String execute()
                 lastUpdatedBy = registration.getLastUpdatedBy();
             }
 
-            locked = dataSetService.isLocked( currentUser, dataSet, period, organisationUnit, attributeOptionCombo,
-                null );
+            locked = dataSetService.getLockStatus( currentUser, dataSet, period, organisationUnit,
+                attributeOptionCombo, null );
+
```
Comment (jimgrace) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java:345]: remove blank line

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java [hunk 7]
```diff
@@ -346,12 +352,16 @@ public String execute()
 
             for ( OrganisationUnit ou : children )
             {
+
```
Comment (jimgrace) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java:355]: remove blank line

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java [hunk 8]
```diff
@@ -346,12 +352,16 @@ public String execute()
...
-                    locked = dataSetService.isLocked( currentUser, dataSet, period, organisationUnit,
-                        attributeOptionCombo, null );
+                    LockStatus lockedInterim;
+                    lockedInterim = dataSetService.getLockStatus( currentUser, dataSet, period, ou,
+                            attributeOptionCombo, null );
+
+                    locked = lockedInterim != LockStatus.OPEN ? lockedInterim : locked;
 
-                    if ( locked )
+                    if ( locked == LockStatus.LOCKED )
```
Comment (jimgrace) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/java/org/hisp/dhis/de/action/GetDataValuesForDataSetAction.java:364]: I believe the following would make this work the way it did before:

```
                    locked = dataSetService.getLockStatus( currentUser, dataSet, period, ou,
                            attributeOptionCombo, null );

                    if ( locked != LockStatus.OPEN )
```

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataset/LockStatus.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.dataset;
+
+public enum LockStatus 
+{
+    LOCKED("locked"),
+    APPROVED("approved"),
+    OPEN(null);
```
Comment (jimgrace) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataset/LockStatus.java:34]: Need a space inside the parentheses on each side of the argument.

**Review by jimgrace — CHANGES_REQUESTED** [review]

**Review by larshelge — APPROVED** [review]

**Review by jimgrace — APPROVED** [review]

**Pull request merged**